contracts/cw20-merkle-airdrop/src/contract.rs

```diff
@@ -0,0 +1,487 @@
+    ConfigResponse, ExecuteMsg, InstantiateMsg, IsClaimedResponse, LatestStageResponse,
+    MerkleRootResponse, MigrateMsg, QueryMsg,
+};
+use cw20::Cw20ExecuteMsg;
+use cw_storage_plus::U8Key;
+use sha3::Digest;
+use std::convert::TryInto;
+
+// Version info, for migration info
+const _CONTRACT_NAME: &str = "crates.io:cw20-merkle-airdrop";
+const _CONTRACT_VERSION: &str = env!("CARGO_PKG_VERSION");
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn instantiate(
+    deps: DepsMut,
+    _env: Env,
+    _info: MessageInfo,
+    msg: InstantiateMsg,
+) -> StdResult<Response> {
+    let config = Config {
+        owner: deps.api.addr_validate(&msg.owner)?,
+        cw20_token_address: deps.api.addr_validate(&msg.cw20_token_address)?,
+    };
+    CONFIG.save(deps.storage, &config)?;
+
+    let stage = 0;
+    STAGE.save(deps.storage, &stage)?;
+
+    Ok(Response::default())
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn execute(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: ExecuteMsg,
+) -> Result<Response, ContractError> {
+    match msg {
+        ExecuteMsg::UpdateConfig { owner } => execute_update_config(deps, env, info, owner),
+        ExecuteMsg::RegisterMerkleRoot { merkle_root } => {
+            execute_register_merkle_root(deps, env, info, merkle_root)
+        }
+        ExecuteMsg::Claim {
+            stage,
+            amount,
+            proof,
+        } => execute_claim(deps, env, info, stage, amount, proof),
+    }
+}
+
+pub fn execute_update_config(
+    deps: DepsMut,
+    _env: Env,
+    info: MessageInfo,
+    owner: Option<String>,
+) -> Result<Response, ContractError> {
+    // authorize owner
+    let cfg = CONFIG.load(deps.storage)?;
+    if info.sender != cfg.owner {
+        return Err(ContractError::Unauthorized {});
+    }
+
+    // validate owner change
+    let new_owner = owner.ok_or(ContractError::InvalidInput {})?;
+    let new_owner = deps.api.addr_validate(&new_owner)?;
+
+    CONFIG.update(deps.storage, |mut exists| -> StdResult<_> {
+        exists.owner = new_owner;
+        Ok(exists)
+    })?;
+
+    Ok(Response {
+        messages: vec![],
+        attributes: vec![attr("action", "update_config")],
+        data: None,
+        events: vec![],
+    })
+}
+
+pub fn execute_register_merkle_root(
+    deps: DepsMut,
+    _env: Env,
+    info: MessageInfo,
+    merkle_root: String,
+) -> Result<Response, ContractError> {
+    // authorize owner
+    let cfg = CONFIG.load(deps.storage)?;
+    if info.sender != cfg.owner {
+        return Err(ContractError::Unauthorized {});
+    }
+
+    // check merkle root length
+    let mut root_buf: [u8; 32] = [0; 32];
+    hex::decode_to_slice(merkle_root.to_string(), &mut root_buf)?;
+
+    let latest_stage: u8 = STAGE.load(deps.storage)?;
+    let stage = latest_stage + 1;
+
+    MERKLE_ROOT.save(deps.storage, U8Key::from(stage), &merkle_root)?;
+    STAGE.save(deps.storage, &stage)?;
+
+    Ok(Response {
```

Comment:
Again, this needs 0.16 Response format as above... just chain the `.add_attribute()` calls